include SHA256 for output files

artoonie opened the pull request:
1. Every file that's written should now use AuditableFile, instead of File. When the file writing is complete, you should call `finalizeAndHash` on it. That will mark it as read-only and add its hash to the audit log.
2. An exception is made for files in the Logger to avoid a circular dependency: how can you write your own hash to your file contents? In this case, we add the hash to the filename.

Alternatively, you could:
```
String hash;
do {
   hash = getHash(logFilePath);
   ReplaceLastLineOfFileWith("The hash of this file is %s".formatted(hash));
} while (getHash(logFilePath) != lastSeenHash)
```
But that will take around 10^60 years to complete by my math.

Resolves #746

tarheel commented on src/main/java/network/brightspots/rcv/AuditableFile.java:
```diff
@@ -0,0 +1,35 @@
+  public void finalizeAndHash() {
+    boolean readOnlySucceeded = setReadOnly();
+    if (!readOnlySucceeded) {
+      Logger.warning("Failed to set file to read-only: %s", getAbsolutePath());
```
Any chance that we want this type of error to trigger `severe` instead of `warning`?

artoonie replied:
I opted for warning since I think there are valid cases where this will fail, though I'm not sure of them. For example, if saving to a mounted Google Drive folder, are you allowed to set a file to readonly?

I haven't tested any of this, but given that the function returns a boolean rather than throwing an exception, I assumed there would be valid, non-severe cases where this fails.

tarheel commented on src/main/java/network/brightspots/rcv/Utils.java:
```diff
@@ -86,4 +91,41 @@ static String getUserName() {
+  static byte[] readFileToByteArray(File file) {
+    byte[] fileBytes = null;
+    try {
+      fileBytes = Files.readAllBytes(file.toPath());
```
Probably not necessary for now, but ideally we would have a way to compute this without reading the entire file into memory at once. Maybe something like what's discussed here would work? https://stackoverflow.com/questions/76013476/java-inputstream-copy-to-and-calculate-hash-at-the-same-time

artoonie replied:
Good call -- I imagine at some point we may decide to hash the input files too, so I've updated this function to read the file in 1kb chunks. Can you take another quick look before I merge?

tarheel reviewed (approved):
> But that will take around 10^60 years to complete by my math.

We could make an M2 processor a requirement for running the tabulator. I bet it could complete this in more like 10^59 years.

artoonie commented on src/main/java/network/brightspots/rcv/AuditableFile.java:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Purpose: Create a file that, on close, is read-only and has its hash added to the audit log.
+ * Design: Overrides the File class
+ * Conditions: Always.
```
`Conditions: Always`? I assume it means "when is this used," since other values include "during tabulation". Since "Always" is used by Logger, I assume it was appropriate to use here too -- i.e. no matter what you do (tabulate, log, launch gui, or do something on CLI), this will be used.

yezr replied:
Is this from some certification requirements?

tarheel commented on src/main/java/network/brightspots/rcv/FileUtils.java:
```diff
@@ -50,6 +56,29 @@ static void createOutputDirectory(String dir) throws UnableToCreateDirectoryExce
     }
   }
 
+  static String getHash(File file) {
+    MessageDigest digest;
+    try {
+      digest = MessageDigest.getInstance("SHA-256");
+    } catch (NoSuchAlgorithmException e) {
+      Logger.severe("Failed to get SHA-256 algorithm");
+      return "[hash not available]";
+    }
+
+    try (InputStream is = Files.newInputStream(file.toPath())) {
+      try (DigestInputStream hashingStream = new DigestInputStream(is, digest)) {
+        while (hashingStream.readNBytes(1024).length > 0) {
+          // Read in 1kb chunks -- don't need to do anything in the body here
+        }
+      }
+    } catch (IOException e) {
+      Logger.severe("Failed to read file: %s", file.getAbsolutePath());
+      return "[hash not available]";
+    }
+
+    return Utils.bytesToHex(digest.digest());
+  }
```
Oh, cool. So this means that the footprint is O(1) instead of O(n)? Also curious why you chose 1024 instead of, say, that number squared or whatever. Not sure it makes any meaningful difference in perf.

artoonie replied:
I assume it doesn't affect performance much either way, though we can revisit if we end up hashing large CVRs and performance becomes a concern.

artoonie commented:
Ready for review with new changes: using a .sha file instead of including the hash in the filename

tarheel commented on src/main/java/network/brightspots/rcv/Logger.java:
```diff
@@ -95,17 +95,19 @@ static void setup() {
       throws IOException {
-    // log file name is: outputFolder + timestamp + log index
-    // FileHandler requires % to be encoded as %%.  %g is the log index
+    // log file name is: outputFolder + timestamp + log index + hash + .log
```
revert?

tarheel commented on src/main/java/network/brightspots/rcv/Utils.java:
```diff
@@ -86,4 +93,27 @@ static String getUserName() {
     return s.trim().split("\\s*\\r?\\n\\s*");
   }
+
+  static byte[] readFileToByteArray(File file) {
```
no longer used, right?

artoonie commented:
1. PR comments fixed
2. Changed `.sha` files to `.hash` and included hash algo in file body
3. Changed signing algo from sha256 to 512

Final review I think! @tarheel

tarheel commented on src/main/java/network/brightspots/rcv/Utils.java:
```diff
@@ -86,4 +93,16 @@ static String getUserName() {
+    StringBuilder hexString = new StringBuilder(2 * hash.length);
+    for (byte b : hash) {
+      String hex = Integer.toHexString(0xff & b);
+      if (hex.length() == 1) {
+        hexString.append('0');
+      }
```
could you explain this part?

artoonie replied:
`toHexString` strips all leading zeros for some reason, so byte 0 would be hex `0` instead of `00`

tarheel replied:
I'm probably just confused here, but why aren't we prepending '0' instead of appending it?

artoonie replied:
We append `0` before we append the remainder of the string, so it's effectively prepending!

tarheel replied:
Sigh. Reading comprehension!
